test/Land/Model/test_physical_bc.jl

```diff
@@ -0,0 +1,420 @@
+    zmax = FT(0)
+    zmin = FT(-1)
+
+    driver_config = ClimateMachine.SingleStackConfiguration(
+        "LandModel",
+        N_poly,
+        nelem_vert,
+        zmax,
+        param_set,
+        m;
+        zmin = zmin,
+        numerical_flux_first_order = CentralNumericalFluxFirstOrder(),
+    )
+
+
+    t0 = FT(0)
+    timeend = FT(150)
+    dt = FT(0.05)
+
+    solver_config = ClimateMachine.SolverConfiguration(
+        t0,
+        timeend,
+        driver_config,
+        ode_dt = dt,
+    )
+    n_outputs = 60
+    mygrid = solver_config.dg.grid
+    every_x_simulation_time = ceil(Int, timeend / n_outputs)
+    state_types = (Prognostic(), Auxiliary(), GradientFlux())
+    dons_arr =
+        Dict[dict_of_nodal_states(solver_config, state_types; interp = true)]
+    time_data = FT[0] # store time data
+
+    # We specify a function which evaluates `every_x_simulation_time` and returns
+    # the state vector, appending the variables we are interested in into
+    # `dons_arr`.
+
+    callback = GenericCallbacks.EveryXSimulationTime(every_x_simulation_time) do
+        dons = dict_of_nodal_states(solver_config, state_types; interp = true)
+        push!(dons_arr, dons)
+        push!(time_data, gettime(solver_config.solver))
+        nothing
+    end
+
+    # # Run the integration
+    ClimateMachine.invoke!(solver_config; user_callbacks = (callback,))
+
+    dons = dict_of_nodal_states(solver_config, state_types; interp = true)
+    push!(dons_arr, dons)
+    push!(time_data, gettime(solver_config.solver))
+
+    # Get z-coordinate
+    z = get_z(solver_config.dg.grid; rm_dupes = true)
+    N = length(dons_arr)
+    computed_surface_flux = [dons_arr[k]["soil.water.K∇h[3]"][end] for k in 1:N]
+    t = time_data
+    prescribed_surface_flux = t -> FT(-1) * FT(3e-8 * sin(pi * 2.0 * t / 300.0))
+    MSE = mean((prescribed_surface_flux.(t) .- computed_surface_flux) .^ 2.0)
+    @test MSE < 1e-7
+end
+
+
+
+@testset "Explicit Dirichlet BC Comparison" begin
+    # This tests that the "if" statement of compute_grad_bc
+    ClimateMachine.init()
+    FT = Float64
+    function init_soil_water!(land, state, aux, localgeo, time)
+        myfloat = eltype(state)
+        state.soil.water.ϑ_l = myfloat(land.soil.water.initialϑ_l(aux))
+        state.soil.water.θ_i = myfloat(land.soil.water.initialθ_i(aux))
+    end
+
+    soil_param_functions = SoilParamFunctions{FT}(
+        porosity = 0.495,
+        Ksat = 0.0443 / (3600 * 100),
+        S_s = 1e-4,
+    )
+    surface_precip_amplitude = FT(-5e-4)
+
+    precip = (t) -> surface_precip_amplitude
+    hydraulics = vanGenuchten{FT}(n = 2.0)
+    function hydrostatic_profile(z, zm, porosity, n, α)
+        myf = eltype(z)
+        m = FT(1 - 1 / n)
+        S = FT((FT(1) + (α * (z - zm))^n)^(-m))
+        return FT(S * porosity)
+    end
+    N_poly = (1, 2)
+    nelem_vert = 50
+
+    # Specify the domain boundaries
+    zmax = FT(0)
+    zmin = FT(-0.35)
+    Δz = FT(0.35 / 50 / 2)
```

Comment:
what should this be in DG? This seems to work? In FV the first point is half a layer down, so this defn makes sense for that. Can we avoid making the grid first and just use a representative value in DG as well? (note: land model likely will use FV in vertical, but whatever we do should work for both).

Reply:
@Zhengyu-Huang do you know what value we would use in DG for the "delta z" in the denominator of the approximated flux (using the value we would use as a dirichlet condition at the surface to estimate the flux that would result)? In DG it's more of an approx because we are using the surface state (state-), so dz is ~0. But using a value typical for the spacing of points (Length of domain/(n_elem*N_poly)) seems to work ok?

Reply:
@Zhengyu-Huang any thoughts on this?

Reply:
For Dg, we should use the first nodal distance, it is ugly, but the vertical nodal distance is
```
using ClimateMachine.Grids
mygrid = solver_config.dg.grid
mygrid.vgeo[1, Grids._x3, 2] - mygrid.vgeo[1, Grids._x3, 1]
```
The first index is the element id, the last index is the quadrature id, the second index is the index for x3 direction

Reply:
Npoly doesnt affect this?

Reply:
We should absolutely write a separate function, in `ClimateMachine.Grids` that computes this. Maybe `lower_Δz(::Grid)` and `upper_Δz(::Grid)`? There is no concern about laterally changing `Δz` here?

Reply:
i dont think we'll change resolution near the surface laterally.. when I tested what Daniel sent, though, it looked like it was just (zmin-zmax)/nelem_vert. If that is the case (at least for cartesian coords) I almost want to punt on writing a function for it because then we end up having to make a grid with an empty model, then create the model, then make the actual grid using the same params...especially if it as simple as the size of a vertical element.

Reply:
in FV the interpretation is more clear (dz = distance from boundary to first point). for DG the first point is at the surface. But perhaps a way to think about it is that in either case, we want to apply the same boundary condition (function of dz) between the two models if they have the same resolution, so we use the same value of dz, even if it is not the distance to the first point anymore?

Reply:
![dg_poly2_convergence](https://user-images.githubusercontent.com/65979205/109057775-0753be00-7697-11eb-9c1b-7b09a16d4910.png)


this is testing the above hypothesis - looks fine? it is using dz = 1/2(zmin-zmax)/nelem_vert as we'd do for FV.

Reply:
Given this, my preference would be to use: `    Δz = abs(FT((zmin-zmax) / nelem_vert / 2))` and to open an issue reminding me that in the future we want a function which calculates this. if i understand correctly, it will only matter if we dont have cartesian/rectangular coordinates, or if we stretch the points. Getting it from the grid is great, but we dont have access to the grid at this point, before we make the model, and I think the solution to that particular issue is pending since it has come up for a number of people.

Reply:
ok having spoken with Daniel we agree this makes sense for DG, but we also want to confirm it. Please see https://github.com/CliMA/ClimateMachine.jl/issues/2058.